megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java

```diff
@@ -0,0 +1,272 @@
+        // it still can't be leveled
+        boolean canLevel = child.needsLeveling() &&
+                !mli.outOfBounds &&
+                !mli.destinationImpassable &&
+                !mli.goingDownTooLow &&
+                !mli.goingUpTooHigh &&
+                !mli.wheeledTankRestriction &&
+                !mli.destinationHasWeakBridge &&
+                !mli.groundTankIntoWater;
+        
+        if((!shortestPathsToCoords.containsKey(child.getFinalCoords()) ||
+                // shorter path to these coordinates
+                (movePathComparator.compare(shortestPathsToCoords.get(child.getFinalCoords()), child) > 0)) &&
+                // legal or needs leveling and not off-board
+                (mli.isLegal() || canLevel) &&
+                // better than existing path to ultimate destination
+                (child.getMpUsed() + child.getLevelingCost() < maximumCost)) {
+            shortestPathsToCoords.put(child.getFinalCoords(), child);
+            children.add(child);
+        }
+    }
+    
+    /**
+     * Comparator implementation useful in comparing how much closer a given path is to the internal
+     * "destination edge" than the other.
+     * @author NickAragua
+     *
+     */
+    private class AStarComparator implements Comparator<BulldozerMovePath> {
+        private Coords destination;
+
+        /**
+         * Constructor - initializes the destination edge.
+         * @param targetRegion Destination edge
+         */
+        public AStarComparator(Coords destination) {
+            this.destination = destination;
+        }
+        
+        /**
+         * compare the first move path to the second
+         * Favors paths that move closer to the destination edge first.
+         * in case of tie, favors paths that cost less MP
+         */
+        public int compare(BulldozerMovePath first, BulldozerMovePath second) {            
+            IBoard board = first.getGame().getBoard();
+            boolean backwards = false;
+            int h1 = first.getFinalCoords().distance(destination)
+                    + ShortestPathFinder.getLevelDiff(first, destination, board)
+                    + ShortestPathFinder.getElevationDiff(first, destination, board, first.getEntity());
+            int h2 = second.getFinalCoords().distance(destination)
+                    + ShortestPathFinder.getLevelDiff(second, destination, board)
+                    + ShortestPathFinder.getElevationDiff(second, destination, board, second.getEntity());
+    
+            int dd = (first.getMpUsed() + first.getLevelingCost() + first.getAdditionalCost() + h1) 
+                    - (second.getMpUsed() + second.getLevelingCost() + second.getAdditionalCost() + h2);
+            
+            // getFacingDiff returns a number between 0 and 3 inclusive. 
+            // if the value diff is larger than 3, then it won't make a difference and we skip calculating it
+            if(Math.abs(dd) < 4) 
```

Comment:
Why not:
```java
if (dd != 0) {
   return dd;
}

dd = ShortestPathFinder.getFacingDiff(first, destination, backwards) - ShortestPathFinder.getFacingDiff(second, destination, backwards);
if (dd != 0) {
    return dd;
 }

return first.getHexesMoved() - second.getHexesMoved();
```
This follows the contract of a multi-step comparator rather than the addition/subtraction of facing diffs. Or is this some sort of optimization to avoid calculating facing diffs?

Reply:
That's basically it, yeah. Calculating facing diffs is expensive because of the arctangents, so I avoid calculating them in situations where it doesn't matter.